NPE in ruby block called concurrently from Java

**Environment Information**
  
* JRuby 9.2.7.0

**Expected Behavior**

No error when calling ruby block concurrently from Java

I have not been able to reproduce this error locally at all but seems to happen sporadically in our testing pipeline. The failing tests are different but a common pattern is passing a ruby block to java which is called concurrently. An example similar to our tests could be:

```ruby
 results = java.util.concurrent.ConcurrentHashMap.new
 calculate_something_in_parallel(input1, input2, -> (id, result) { results[id] = result })
```
The stack trace then indicates that it fails when calling the lambda. 

**Actual Behavior**

Returns an NPE with the following stack trace:

```
java.lang.NullPointerException 
     org.jruby.runtime.InterpretedIRBlockBody.commonYieldPath(InterpretedIRBlockBody.java:127)
     org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:79)
     org.jruby.runtime.Block.call(Block.java:124)
     org.jruby.RubyProc.call(RubyProc.java:295)
     org.jruby.RubyProc.call(RubyProc.java:274)
     org.jruby.RubyProc.call(RubyProc.java:270)
     org.jruby.javasupport.Java$ProcToInterface.callProc(Java.java:1131)
     org.jruby.javasupport.Java$ProcToInterface.access$300(Java.java:1108)
     org.jruby.javasupport.Java$ProcToInterface$ConcreteMethod.call(Java.java:1169)
     org.jruby.gen.InterfaceImpl630006984.consume(org/jruby/gen/InterfaceImpl630006984.gen:13)
```

@headius I found a way that reproduces this bug very rarely and I also found another sporadic error in the process. I attach in a zip the files for this reproduction. To run them: `javac Repo.java` and then `jruby --dev script.rb`

What I found was that it seems that the bug only reproduces using the "--dev" flag. Tested this flag with the snippet you pasted here and it did not seem to work for me (left it running ~15min)

The new error I found running my script has the following stacktrace and it seems to appear commonly:

> Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
> 	at org.jruby.runtime.scope.ManyVarsDynamicScope.setValueDepthZeroVoid(ManyVarsDynamicScope.java:183)
> 	at org.jruby.runtime.scope.ManyVarsDynamicScope.setValueVoid(ManyVarsDynamicScope.java:174)
> 	at org.jruby.ir.interpreter.InterpreterEngine.setResult(InterpreterEngine.java:572)
> 	at org.jruby.ir.interpreter.InterpreterEngine.receiveArg(InterpreterEngine.java:273)
> 	at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:68)
> 	at org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(Interpreter.java:116)
> 	at org.jruby.runtime.InterpretedIRBlockBody.callDirect(InterpretedIRBlockBody.java:102)
> 	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)
> 	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:52)
> 	at org.jruby.runtime.Block.call(Block.java:139)
> 	at org.jruby.RubyProc.call(RubyProc.java:318)
> 	at org.jruby.javasupport.Java$ProcToInterface.callProc(Java.java:1137)
> 	at org.jruby.javasupport.Java$ProcToInterface.access$300(Java.java:1114)
> 	at org.jruby.javasupport.Java$ProcToInterface$ConcreteMethod.call(Java.java:1175)
> 	at org.jruby.gen.InterfaceImpl714518509.accept(org/jruby/gen/InterfaceImpl714518509.gen:13)
> 	at Repro.lambda$pimpam$0(Repro.java:18)
> 	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1626)

I tested with 9.2.7.0 and 9.2.11.0, curiously enough I was only able to get the NPE when running with 9.2.7.0 but might just be a question of leaving 9.2.11.0 running longer.

[reproduction.zip](https://github.com/jruby/jruby/files/4786202/reproduction.zip)

--- a/core/src/main/java/org/jruby/runtime/InterpretedIRBlockBody.java
+++ b/core/src/main/java/org/jruby/runtime/InterpretedIRBlockBody.java
@@ -76,7 +76,6 @@ public class InterpretedIRBlockBody extends IRBlockBody implements Compilable<In
 
             ic = closure.getInterpreterContext();
             interpreterContext = ic;
-            fullInterpreterContext = interpreterContext;
         }
         return ic;
     }
@@ -93,19 +92,19 @@ public class InterpretedIRBlockBody extends IRBlockBody implements Compilable<In
 
     @Override
     public boolean canCallDirect() {
-        return interpreterContext != null && interpreterContext.hasExplicitCallProtocol();
+        return fullInterpreterContext != null && fullInterpreterContext.hasExplicitCallProtocol();
     }
 
     @Override
     protected IRubyObject callDirect(ThreadContext context, Block block, IRubyObject[] args, Block blockArg) {
-        InterpreterContext ic = ensureInstrsReady(); // so we get debugging output
-        return Interpreter.INTERPRET_BLOCK(context, block, null, ic, args, block.getBinding().getMethod(), blockArg);
+        ensureInstrsReady(); // so we get debugging output
+        return Interpreter.INTERPRET_BLOCK(context, block, null, fullInterpreterContext, args, block.getBinding().getMethod(), blockArg);
     }
 
     @Override
     protected IRubyObject yieldDirect(ThreadContext context, Block block, IRubyObject[] args, IRubyObject self) {
-        InterpreterContext ic = ensureInstrsReady(); // so we get debugging output
-        return Interpreter.INTERPRET_BLOCK(context, block, self, ic, args, block.getBinding().getMethod(), Block.NULL_BLOCK);
+        ensureInstrsReady(); // so we get debugging output
+        return Interpreter.INTERPRET_BLOCK(context, block, self, fullInterpreterContext, args, block.getBinding().getMethod(), Block.NULL_BLOCK);
     }
 
     @Override
@@ -113,11 +112,6 @@ public class InterpretedIRBlockBody extends IRBlockBody implements Compilable<In
         if (callCount >= 0) promoteToFullBuild(context);
 
         InterpreterContext ic = ensureInstrsReady();
-
-        // Update interpreter context for next time this block is executed
-        // This ensures that if we had determined canCallDirect() is false
-        // based on the old IC, we continue to execute with it.
-        interpreterContext = fullInterpreterContext;
 
         Binding binding = block.getBinding();
         Visibility oldVis = binding.getFrame().getVisibility();

